File: functions/src/notification_preferences.ts
import * as admin from 'firebase-admin';

export const NOTIFICATION_PREFS_DOC_ID = 'notifications';
export const NOTIFICATION_PREFS_COLLECTION = 'settings';

export type NotificationPrefs = {
  pushEnabled: boolean;
  activity: boolean;
  postsFromFollowing: boolean;
  live: boolean;
  subscriptions: boolean;
  recommended: boolean;
};

const DEFAULT_PREFS: NotificationPrefs = {
  pushEnabled: true,
  activity: true,
  postsFromFollowing: true,
  live: true,
  subscriptions: false,
  recommended: false,
};

function readBool(value: unknown, fallback: boolean): boolean {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'number') return value !== 0;
  if (typeof value === 'string') {
    const t = value.trim().toLowerCase();
    if (t === 'true' || t === '1') return true;
    if (t === 'false' || t === '0') return false;
  }
  return fallback;
}

export function parseNotificationPrefs(
  data: admin.firestore.DocumentData | undefined,
): NotificationPrefs {
  if (!data) return { ...DEFAULT_PREFS };
  return {
    pushEnabled: readBool(data.pushEnabled, DEFAULT_PREFS.pushEnabled),
    activity: readBool(data.activity, DEFAULT_PREFS.activity),
    postsFromFollowing: readBool(
      data.postsFromFollowing,
      DEFAULT_PREFS.postsFromFollowing,
    ),
    live: readBool(data.live, DEFAULT_PREFS.live),
    subscriptions: readBool(data.subscriptions, DEFAULT_PREFS.subscriptions),
    recommended: readBool(data.recommended, DEFAULT_PREFS.recommended),
  };
}

export async function loadNotificationPrefs(
  db: admin.firestore.Firestore,
  uid: string,
): Promise<NotificationPrefs> {
  if (!uid.trim()) return { ...DEFAULT_PREFS };
  const snap = await db
    .collection('users')
    .doc(uid.trim())
    .collection(NOTIFICATION_PREFS_COLLECTION)
    .doc(NOTIFICATION_PREFS_DOC_ID)
    .get();
  return parseNotificationPrefs(snap.data());
}

/** Returns false when push should be suppressed for this notification type. */
export function shouldSendPushForType(
  prefs: NotificationPrefs,
  rawType: string,
): boolean {
  if (!prefs.pushEnabled) return false;
  const type = rawType.trim().toLowerCase();
  if (!type) return prefs.activity;

  const activityTypes = new Set([
    'like',
    'comment',
    'share',
    'follow',
    'followrequest',
    'follow_request',
    'followrequestaccepted',
    'follow_request_accepted',
  ]);
  if (activityTypes.has(type)) return prefs.activity;

  if (type === 'subscribe' || type === 'subscription') {
    return prefs.subscriptions;
  }

  const postTypes = new Set(['post', 'new_post', 'newpost']);
  if (postTypes.has(type)) return prefs.postsFromFollowing;

  const liveTypes = new Set(['live', 'live_start', 'livestream', 'live_stream']);
  if (liveTypes.has(type)) return prefs.live;

  const recommendedTypes = new Set([
    'recommended',
    'recommended_content',
    'marketing',
  ]);
  if (recommendedTypes.has(type)) return prefs.recommended;

  return prefs.activity;
}

File: lib/core/models/notification_preferences.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// User-controlled push/in-app notification categories.
/// Stored at `users/{uid}/settings/notifications`.
class NotificationPreferences {
  const NotificationPreferences({
    this.pushEnabled = true,
    this.activity = true,
    this.postsFromFollowing = true,
    this.live = true,
    this.subscriptions = false,
    this.recommended = false,
  });

  final bool pushEnabled;
  final bool activity;
  final bool postsFromFollowing;
  final bool live;
  final bool subscriptions;
  final bool recommended;

  static const String firestoreDocId = 'notifications';
  static const String collectionName = 'settings';

  static NotificationPreferences fromMap(Map<String, dynamic>? data) {
    if (data == null || data.isEmpty) {
      return const NotificationPreferences();
    }
    return NotificationPreferences(
      pushEnabled: _readBool(data['pushEnabled'], fallback: true),
      activity: _readBool(data['activity'], fallback: true),
      postsFromFollowing: _readBool(data['postsFromFollowing'], fallback: true),
      live: _readBool(data['live'], fallback: true),
      subscriptions: _readBool(data['subscriptions'], fallback: false),
      recommended: _readBool(data['recommended'], fallback: false),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'pushEnabled': pushEnabled,
      'activity': activity,
      'postsFromFollowing': postsFromFollowing,
      'live': live,
      'subscriptions': subscriptions,
      'recommended': recommended,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }

  NotificationPreferences copyWith({
    bool? pushEnabled,
    bool? activity,
    bool? postsFromFollowing,
    bool? live,
    bool? subscriptions,
    bool? recommended,
  }) {
    return NotificationPreferences(
      pushEnabled: pushEnabled ?? this.pushEnabled,
      activity: activity ?? this.activity,
      postsFromFollowing: postsFromFollowing ?? this.postsFromFollowing,
      live: live ?? this.live,
      subscriptions: subscriptions ?? this.subscriptions,
      recommended: recommended ?? this.recommended,
    );
  }

  /// Whether a Firestore notification [type] should surface as push/in-app alert.
  bool allowsCategoryForType(String rawType) {
    if (!pushEnabled) return false;
    final type = rawType.trim().toLowerCase();
    if (type.isEmpty) return activity;

    const activityTypes = {
      'like',
      'comment',
      'share',
      'follow',
      'followrequest',
      'follow_request',
      'followrequestaccepted',
      'follow_request_accepted',
    };
    if (activityTypes.contains(type)) return activity;

    if (type == 'subscribe' || type == 'subscription') return subscriptions;

    const postTypes = {'post', 'new_post', 'newpost'};
    if (postTypes.contains(type)) return postsFromFollowing;

    const liveTypes = {'live', 'live_start', 'livestream', 'live_stream'};
    if (liveTypes.contains(type)) return live;

    const recommendedTypes = {
      'recommended',
      'recommended_content',
      'marketing',
    };
    if (recommendedTypes.contains(type)) return recommended;

    return activity;
  }

  static bool _readBool(dynamic value, {required bool fallback}) {
    if (value is bool) return value;
    if (value is num) return value != 0;
    if (value is String) {
      final t = value.trim().toLowerCase();
      if (t == 'true' || t == '1') return true;
      if (t == 'false' || t == '0') return false;
    }
    return fallback;
  }
}
